shared/task.rs

```
@@ -1,89 +1,404 @@
-        return self.status == TaskStatus::Queued || self.status == TaskStatus::Stashed;
+        match &self.state {
+            TaskState::Waiting { status: Waiting::Queued, .. } => true,
+            _ => false,
+        }
+    }
+}
+
+// Get general info about task in it's context
+pub enum Startability {
+    Ready,
+    Waiting(Duration),
+    Dependencies(Vec<usize>),
+    DependenciesFailure(usize),
+    Stashed,
+    Locked,
+
+    Unknown,
+}
+
+pub enum GeneralState {
+    Healthy,
+    Failed,
+    Paused,
+    Waiting,
+}
+
+impl Task {
+    pub fn can_be_started(&self, state: &State) -> bool {
+        match self.start_info(state) {
+            Startability::Ready => true,
+            _ => false,
+        }
+    }
+
+    pub fn start_info(&self, state: &State) -> Startability {
+        match &self.state {
+            TaskState::Waiting {
+                locked: true, ..
+            } => Startability::Locked,
+
+            TaskState::Waiting {
+                status: Waiting::Stashed,
+                ..
+            } => Startability::Stashed,
+
+            TaskState::Waiting {
+                status: Waiting::Queued,
+                enqueue_at: Some(time),
+                ..
+            } if time > &Local::now() => {
+                Startability::Waiting(*time - Local::now())
+            }
+
+            TaskState::Waiting {
+                status: Waiting::Queued,
+                dependencies,
+                ..
+            } => {
+                let dep_not_done: Vec<_> = dependencies
+                    .iter()
+                    .flat_map(|id| state.get_task(*id))
+                    .filter(|task| !task.is_done())
+                    .collect();
+
+                if !dep_not_done.is_empty() {
+                    let failed = dep_not_done.iter().find(|task| task.is_errored());
+
+                    if let Some(failed_task) = failed {
+                        Startability::DependenciesFailure(failed_task.id)
```

A tasks with failed dependencies should fail hard.
As far as I understand, right now those tasks will still be in internal `Waiting` state, but displayed as failed to the outside?